### ISSUE TITLE
Add IRouteDiagnosticsMetadata

### ISSUE THREAD

**Issue by JamesNK** [dotnet/aspnetcore#47492]
## Background and Motivation

The Open Telemetry [HTTP server semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/4fa8a0dcfe8b3f8413f8626067c40fca2c46613d/specification/metrics/semantic_conventions/http-metrics.md#http-server) records required and optional data to include with metrics measurements. A recommended tag is the route match by the HTTP request. For example, an HTTP request to `app.MapGet("product/{id}", () => ...)` would have the route `product/{id}`.

The route is available in ASP.NET Core on the matched endpoint in the [RouteEndpoint.RoutePattern](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.routing.routeendpoint.routepattern?view=aspnetcore-7.0#microsoft-aspnetcore-routing-routeendpoint-routepattern). However, HTTP metrics are run in `Microsoft.AspNetCore.Hosting`. That code doesn't have access to `RouteEndpoint`, which is in `Microsoft.AspNetCore.Routing`, because of layering.

Two possible solutions:

* Move `RouteEndpoint`, `RoutePattern`, and all its dependencies to `Microsoft.AspNetCore.Http.Abstractions` and type forward. Then hosting can check if the matched endpoint is a `RouteEndpoint` and access the route. However, the route pattern is quite a lot of code to move.
* Or, add new diagnostics metadata to `Microsoft.AspNetCore.Http.Abstractions`. The metadata has a property with a string version of the pattern. Hosting diagnostics can then get the metadata from the endpoint, and use whatever text value is present.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Http.Metadata;

+ public interface IRouteDiagnosticsMetadata
+ {
+     string Route { get; }
+ }
```

## Usage Examples

```csharp
if (context.MetricsEnabled)
{
    var route = httpContext.GetEndpoint()?.Metadata.GetMetadata<IRouteDiagnosticsMetadata>()?.Route;

    _metrics.RequestEnd(
        httpContext.Request.Protocol,
        httpContext.Request.Scheme,
        httpContext.Request.Method,
        httpContext.Request.Host,
        route,
        httpContext.Response.StatusCode,
        startTimestamp,
        currentTimestamp);
}
```

## Alternative Designs

See description.

**Comment by halter73** [comment]
API Review Note:

- Would we ever consider moving `RouteEndpoint` and `RoutePattern` down into Http.Abstractions.dll?
  - `RoutePattern` might be too big when all we need is the string for stuff like diagnostics and logging.
- [Microsoft.AspNetCore.Http.Metadata](https://learn.microsoft.com/en-nz/dotnet/api/microsoft.aspnetcore.http.metadata?view=aspnetcore-670) or Microsoft.AspNetCore.Routing?
- Can we just have Routing.dll add metadata with the actual `RoutePattern`? If you have a reference to routing, you can just cast the `Endpoint` to a `RouteEndpoint`.
  - `var route = ((RouteEndpoint)httpContext.GetEndpoint()).RoutePattern`
- Do we like the name `Route` for the property? What about `RoutePattern`?
  - We don't want to get peoples hopes up that they're getting the actual `RoutePattern` type.
- What about the `IRouteDiagnosticsMetadata` type name? There could be a lot of "Diagnostics". Would we add more to it in the future? Probably not.

API Approved!

```diff
// Microsoft.AspNetCore.Http.Abstractions.dll

namespace Microsoft.AspNetCore.Http.Metadata;

+ public interface IRouteDiagnosticsMetadata
+ {
+     string Route { get; }
+ }
```

**Comment by amcasey** [comment]
@JamesNK was this fixed by #46834?
